jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/AHUForestIsomorphismInspector.java

```diff
@@ -0,0 +1,235 @@
+ * </p>
+ *
+ * <p>
+ *     Note: This implementation requires the input graphs to be modifiable and to have valid vertex suppliers
+ *     (see {@link Graph#getVertexSupplier()}).
+ * </p>
+ *
+ * @param <V> the type of the vertices
+ * @param <E> the type of the edges
+ *
+ * @author Alexandru Valeanu
+ */
+public class AHUForestIsomorphismInspector<V, E> {
+    private final Graph<V, E> forest1;
+    private final Graph<V, E> forest2;
+
+    private Set<V> roots1;
+    private Set<V> roots2;
+
+    private boolean computed = false;
+    private IsomorphicTreeMapping<V, E> isomorphicMapping = null;
+
+    /**
+     * Construct a new AHU rooted tree isomorphism inspector.
+     *
+     * Note: The constructor does NOT check if the input trees are valid.
+     *
+     * @param tree1 the first rooted tree
+     * @param root1 the root of the first tree
+     * @param tree2 the second rooted tree
+     * @param root2 the root of the second tree
+     */
+    public AHUForestIsomorphismInspector(Graph<V, E> tree1, V root1, Graph<V, E> tree2, V root2){
+        this(tree1, Collections.singleton(Objects.requireNonNull(root1, "root cannot be null")),
+                tree2, Collections.singleton(Objects.requireNonNull(root2, "root cannot be null")));
+    }
+
+    /**
+     * Construct a new AHU rooted forest isomorphism inspector.
+     *
+     * Note: The constructor does NOT check if the input forests are valid.
+     *
+     * @param forest1 the first rooted forest
+     * @param roots1 the roots of the first forest
+     * @param forest2 the second rooted forest
+     * @param roots2 the roots of the second forest
+     */
+    public AHUForestIsomorphismInspector(Graph<V, E> forest1, Set<V> roots1, Graph<V, E> forest2, Set<V> roots2){
+        this.forest1 = Objects.requireNonNull(forest1, "input forest cannot be null");
+        this.forest2 = Objects.requireNonNull(forest2, "input forest cannot be null");
+
+        this.roots1 = Objects.requireNonNull(roots1, "set of roots cannot be null");
+        this.roots2 = Objects.requireNonNull(roots2, "set of roots cannot be null");
+
+        if (roots1.isEmpty()){
+            throw new IllegalArgumentException("root set cannot be empty");
+        }
+
+        if (roots2.isEmpty()){
+            throw new IllegalArgumentException("root set cannot be empty");
+        }
+
+        if (!forest1.vertexSet().containsAll(roots1)){
+            throw new IllegalArgumentException("root not contained in forest");
+        }
+
+        if (!forest2.vertexSet().containsAll(roots2)){
+            throw new IllegalArgumentException("root not contained in forest");
+        }
+    }
+
+    private V getFreshVertex(Graph<V, E> graph){
+        Supplier<V> supplier = graph.getVertexSupplier();
+
+        if (Objects.isNull(supplier))
+            throw new IllegalArgumentException("vertex supplier cannot be null");
+
+        while (true){
```

Comment:
It is since the graph might not have been constructed using the supplier.

Reply:
I see what you mean, but that already violates our intended usage for suppliers.  For example, if I create a graph with integers as vertices, with a sequential int supplier which starts from 1, and then add vertices 1, 2, 3 explictl via addVertex(V), and then I try to call the no-arg addVertex(), I'll get an exception.  @d-michail I'm not sure we've documented this caveat?

So in your case, you can just assert instead of looping.